### PR TITLE
Simple Send v1 with Memo field

### PR DESCRIPTION
**Pull request by zathras-crypto**
@CraigSellars has requested a priority addition to simple send, to allow a note/memo to be attached with the transaction.

This PR serves as proof of concept code for the addition and as a placeholder for discussion of the change.  It is not intended to be merged as-is.

`omni_send` has been adjusted to allow for an optional memo string.  If text for the memo field is supplied the payload will be created v1 with the memo, if none is supplied the payload will be created v0 with no memo.  

@dexX7 I reordered the params on `omni_send` so a memo could be sent without having to specify manual redeem address and reference amounts - it feels better this way but does change the parameter order which I usually don't like doing (can break backwards compat) - conversely I see a memo being used probably more often that manual redeem addresses/reference amounts.  What do you think?

I haven't added feature activation for this one, I'm not 100% sure how we should handle it.  Basically because the memo is appended to the end of the payload in this proposed v1 simple send, we could actually skip incrementing the transaction version.  Earlier versions of Omni Core only look at the first 16 bytes of a simple send and would decode it fine (just without the memo).  So it shouldn't be consensus breaking on unsupported clients if we added memos to v0 transactions, though I'm also not really a fan of that idea.  Currently I built the PR with separate versions, and if we keep it this way I'll need to add activation for the new version.

A regular simple send gets sent as v0

```
$ src/omnicore-cli --testnet omni_send mkS9rGbmSefsWmBkCYGJpzcrJbcaziWVLK mfv4NUBwzFf81zxpXHTJNonYXKZTZi3VuZ 1 0.1
3c7caa789c8bdbee7c8f914714b0b22130cf07dfe869843ba1c45a7e2ae36dff
$ src/omnicore-cli --testnet omni_gettransaction 3c7caa789c8bdbee7c8f914714b0b22130cf07dfe869843ba1c45a7e2ae36dff | grep version
    "version" : 0,
```

A simple send with a memo gets sent as v1

```
$ src/omnicore-cli --testnet omni_send mkS9rGbmSefsWmBkCYGJpzcrJbcaziWVLK mfv4NUBwzFf81zxpXHTJNonYXKZTZi3VuZ 1 2.1 "Test Memo"
f64599f8a87f9f036142d5b00dea4669800a44602952ee1a5123ad291b68e29b
$ src/omnicore-cli --testnet omni_gettransaction f64599f8a87f9f036142d5b00dea4669800a44602952ee1a5123ad291b68e29b | grep version
    "version" : 1,
```

Decoding is pretty straight forward:

```
parseTransaction(block=577671, 2015-10-13 06:51:23 idx= 8); txid: f64599f8a87f9f036142d5b00dea4669800a44602952ee1a5123ad291b68e29b
2015-10-13 06:53:32     ------------------------------
2015-10-13 06:53:32              version: 1, class C
2015-10-13 06:53:32                 type: 0 (Simple Send)
2015-10-13 06:53:32             property: 1 (OMNI)
2015-10-13 06:53:32                value: 2.10000000
2015-10-13 06:53:32                 memo: Test Memo
2015-10-13 06:53:32 update_tally_map(mkS9rGbmSefsWmBkCYGJpzcrJbcaziWVLK, 1=0x1, -210000000, ttype=0): before=1680000000, after=1470000000
2015-10-13 06:53:32 update_tally_map(mfv4NUBwzFf81zxpXHTJNonYXKZTZi3VuZ, 1=0x1, +210000000, ttype=0): before=120000000, after=330000000
2015-10-13 06:53:32 recordTX(f64599f8a87f9f036142d5b00dea4669800a44602952ee1a5123ad291b68e29b, valid=YES, block= 577671, type= 0, value= 210000000)

{
    "txid" : "f64599f8a87f9f036142d5b00dea4669800a44602952ee1a5123ad291b68e29b",
    "fee" : "0.00027853",
    "sendingaddress" : "mkS9rGbmSefsWmBkCYGJpzcrJbcaziWVLK",
    "referenceaddress" : "mfv4NUBwzFf81zxpXHTJNonYXKZTZi3VuZ",
    "ismine" : true,
    "version" : 1,
    "type_int" : 0,
    "type" : "Simple Send",
    "propertyid" : 1,
    "divisible" : true,
    "amount" : "2.10000000",
    "memo" : "Test Memo",
    "valid" : true,
    "blockhash" : "000000000db09a900cfbb4a9391262e630380821943e55909c300d616094ab3d",
    "blocktime" : 1444719083,
    "block" : 577671,
    "confirmations" : 1
}
```

Thoughts on approach and concept code?

Thanks
Z

### CODE REVIEW

**Comment by dexX7** [comment]
I don't think the version bump is necessary at all.

Messages are not consensus relevant, and might just be stuffed anywhere, i.e.:

[simple send] [message]

and:

**Comment by dexX7** [comment]
I don't think the version bump is necessary at all.

Messages are not consensus relevant, and might just be stuffed anywhere, e.g.:

[simple send] [message]

If it comes down to identification, then the message could potentially be tagged, or based on the payload size (class C) the memo might be identified.

I'd prefer, if this doesn't mess with potential multi-transactions and virtual destinations, e.g.:

[simple send] [destination] [simple send] [destination] ...

But since there is not yet a solid concept, I probably can't insist to reserve the space. :(

File: src/omnicore/tx.cpp [hunk 1]
```diff
@@ -199,6 +199,19 @@ bool CMPTransaction::interpret_SimpleSend()
...
+    if (version == MP_TX_PKT_V1 && pkt_size > 16) {
+        const char* p = 16 + (char*) &pkt;
+        std::string memostr(p);
+        memcpy(memo, memostr.c_str(), std::min(memostr.length(), sizeof(memo)-1));
+        if ((!rpcOnly && msc_debug_packets) || msc_debug_packets_readonly) {
+            PrintToLog("\t            memo: %s\n", memo);
+        }
+        if (isOverrun(p)) {
+            PrintToLog("%s(): rejected: malformed string value(s)\n", __func__);
+            return false;
```
Comment (dexX7) [inline src/omnicore/tx.cpp]: This actually makes the message consensus relevant, and might not work at all. Consider a class B encoded transaction with:

[simple send] [random bytes]

... which has no null terminator.

**Comment by dexX7** [comment]
On a related note: might be useful to have a more generalized approach. What if @CraigSellars wants messages attached to STO tomorrow..?

Sorry, shouldn't sound snippy. :)

**Comment by zathras-crypto** [comment]
> ... which may not have a null terminator.

A Class B transaction with random bytes after the simple send and no null terminator would not be a valid version 1 simple send.  v0 requires 4 fields, v1 requires a 5th memo field.

> Messages are not consensus relevant, and might just be stuffed anywhere, e.g.:

So you would propose that a message be appended to any payload, rather than being part of a specific transaction payload?

ie ```[simple send] [memo]``` vs ```[simple send inc memo]```?

I agree this would provide more flexibility, but feels less robust than a field within a transaction payload (like SP strings etc).

> I'd prefer, if this doesn't mess with potential multi-transactions and virtual destinations, e.g.:

Well, I'd originally intended this to be a field just the same as Grant and Revokes (as per spec, memo not implemented in code).  So in that manner it wouldn't mess with multi-transactions and virtual destinations:

```[simple send v1 inc memo] [destination] [simple send v1 inc memo] [destination]``` ...

> Edit: hehe, there is no code to read "grant/revoke" messages. ;)

Nope, I always thought such a low hanging thing should get done, but it's never been requested hehe

> On a related note: might be useful to have a more generalized approach

Well, this is kind of the same discussion - should a memo/message be part of the payload or appended to it?  I think making it a field within a payload provides more granular control.  Eg in your multi-transaction example having the memo as a field within the payload allows each send to have it's own memo.

**Comment by dexX7** [comment]
Just to provide some quick background, why I mentioned the virtual destinations/multi payloads (sort of out of the blue):

1. extra messages add an extra meaning, which is kinda similar to an extra payload [...]
2. more important: the minrelaytxfee of Bitcoin Core was raised, and is going to be floating at some point soon. This probably breaks our assumption of "dust thresholds", and the less need we have for dust-ish outputs (e.g. references), the better

> So you would propose that a message be appended to any payload, rather than being part of a specific transaction payload?

Let me rephrase: in the longer run my goal would be to get rid of all non-critical fields, such as messages (or property descriptions, ...), and make them non-consensus-relevant, and adding an extra field to simple sends for messages goes in the opposite direction, so I'm trying to find an alternative. :)

> I agree this would provide more flexibility, but feels less robust than a field within a transaction payload (like SP strings etc).

Can you clarify what you are referring to with "less robust"? I guess the tricky part would be to differentiate "message" from "junk".

>  I think making it a field within a payload provides more granular control. Eg in your multi-transaction example having the memo as a field within the payload allows each send to have it's own memo.

I see where you're going.

I think there are a few options, e.g. seperated payloads, each with their own version/type, simply appending data etc., but I have a hard time to imagine what would be a good way.

From another perspective:

After the parsing we have blob of bytes, which is our payload, and potentially junk. We currently extract meaning from the payload based on the transaction prefix (i.e. version, type) during the interpretation. Destinations are currently part of the parsing, and not part of the interpretation. Do you think it's feasible to continue that path? It would *really* be awesome, if we could, because then we'd just have to tweak the parsing, but basically nothing else (well, probably it's more, hehe, ... but let's keep it simple).

For multi-transactions a similar question arrises, but it's probably more complicated, because if there were multiple payloads, then we'd have to go through the following steps (i.e. interpretation) multiple times, in one way or another.

Let's say, `parseTransaction`, in a simplified version, yields:
```
{sender, receiver, payload}
```

If there were multiple payloads, then ideally we'd get a collection after the parsing, e.g.:
```
[{sender, receiver, payload}, {sender, receiver, payload}, ...]
```
... which could then be passed further. Let's ignore how transactions/actions are identified (e.g. via txid) for now, as this is probably a broader topic. Even with multi-transactions, we probably have only one sender, but let's ignore this, too. This data of `{s, r, p}` is basically the essence of what we're looking for.

If we assume the initial parsing via `parseTransaction` remains the same, and if we assume that the follow up steps also remain the same (i.e. interpretation), then we'd probably have to insert a new step, to convert:
```
{sender, receiver, payload} -> [{sender, receiver, payload}, {sender, receiver, payload} ...]
```
... and basically split the payload somehow, without really knowing what the payload subparts mean. Do you think this is feasible?

I'm not sure, if this makes sense at all, but I'm just brainstorming.

But where I'm going:

All this doesn't really seem to have a relation to memos, and it looks like:

1. an extra message may still be "hidden" in junk after actual message in the payload
2. an extra message could become part of the payload
3. an extra message could become a seperate transaction (in a multi-transaction)

Now I *do* have to admit that just bumping the version and adding a memo as new field seems like an easy way, but I really don't like it. :(

Many words, but I guess we're not really making progress. Sorry! :)

**Comment by zathras-crypto** [comment]
Apologies, thought I had posted but can't find my comment :(

I spoke with @CraigSellars a few days ago about the use case for this, and it is essentially "tagging" a specific token.  So a message would be sent along with the token in a simple send, and this would "tag" the token with a particular attribute. 

Relevant discussion:
```
[4:24:29 PM] zathrasc: Hey Craig, can you please help me understand your use case and give me a bit of a walkthrough?  I built some proof of concept code for messaging in simple send, and there is some debate as to how to approach it - I'm trying to get it ready for the 0.0.10 RC we'll be doing in the coming days, but if I had some further info on use case I might be able to steer the discussion a bit
[5:02:17 PM] Craig Sellars: The idea is that I can hold ten of some particular object/asset, but the "state" of each of these objects may differ. The "state" of the object is defined in the message during the send of the object.
[5:02:42 PM] Craig Sellars: Subsequent to that, the message will tell "which" of the objects was sent
[5:02:56 PM] Craig Sellars: So I'm in Shanghai, talking to Vitalik.
[5:03:20 PM] zathrasc: nice!
[5:03:47 PM] zathrasc: what if the object is then on-sent by the recipient?
```

I think in a nutshell messaging gives people the ability to build their own applications on top of basic transactions like simple send by including their own data.  Omnicore would simply provide the message that was included with a transaction, but how that message is used by integrators could be for all kinds of funky things.

I think this is the one issue that's kind of holding up the RC, because I know @CraigSellars is really keen to see this in 0.0.10 so I think we should try and set a direction (even if it's more complex than my proof-of-concept code) and start working towards it.

So I guess Q numero uno is "do we want to do messaging on a protocol, or per transaction basis?"

> Let me rephrase: in the longer run my goal would be to get rid of all non-critical fields, such as messages (or property descriptions, ...), and make them non-consensus-relevant, and adding an extra field to simple sends for messages goes in the opposite direction, so I'm trying to find an alternative. :)

This makes complete sense - though I have a slightly differing viewpoint - why "get rid" of all non-critical fields?  We just don't make consensus reliant on those fields - in most cases I think we're already going in that direction right?  I mean if someone where to build a new client that interpreted the strings in property descriptions differently, that isn't a problem from a consensus point of view.  Isn't the "low hanging fruit" approach simply to denominate which fields are consensus relevant (and thus must be accurate) in the spec?  Anything else is then essentially a "nice to have" from a client implementation PoV?

> Can you clarify what you are referring to with "less robust"? I guess the tricky part would be to differentiate "message" from "junk".

Exactly that - I like specifics, and I like structure :)  Just reading some data after the packet that may be random, may be padding, or may be a message and then guessing at which is a little rough, which is why I felt a defined field would be more robust.  

> 1 an extra message may still be "hidden" in junk after actual message in the payload
> 2 an extra message could become part of the payload
> 3 an extra message could become a seperate transaction (in a multi-transaction)

I'm kind of fond of 2, because I'm a believer that the "payload" is the master message, and all data pertaining to the transaction should fall within it.  It keeps the payload atomic, which I feel facilitates easier future things like multi-destination and multi-payload.

Alternatively, we could use magic bytes to denominate purpose, eg:
```{P}PayloadAction{R}Hash160ofAddress{M}Message```
Then they're easily chained also:
```{P}PayloadAction{R}Hash160ofAddress{M}Message{P}PayloadAction2{R}Hash160ofAddress2{M}Message2```

It does waste a little space though, but gives explicit meaning to each section of data.

> Now I do have to admit that just bumping the version and adding a memo as new field seems like an easy way, but I really don't like it. :(

That's absolutely fine hehe, we just need to find a solution that we're all comfortable with :)

**Comment by dexX7** [comment]
Just as an idea, what if we use one bit of the version/type to indicate extra fields, which is masked during the parsing?

Say for example, if the actual type is `0x0000`, then we'd use `0x8000` to indicate an appended message, which is handled in an extra step, and during (old) parsing, we ignore the first bit?

Or we modify the assumption that the prefix is always `[version] [type]`, to something like `[packet-info] [version] [type]`, where the version field is split into 2x 8 bit fields, instead of 1x 16 bit field?

I'm not sure, if we gain much by such an approach, but currently we have quite a bit of space unused.

**Comment by dexX7** [comment]
As a side note: if we accomplish that the message field is not consensus critical, than it's not a strictly required for 0.0.10.

**Comment by dexX7** [comment]
Regarding multi packets with explicit prefix: if we change the meaning of the first field (i.e. version), then we may introduce a packet sequence number, starting at `0x00`. This preserves the original meaning (as all versions currently start with `0x00??`, basically ending up with something like:

```
[0x00] [packet 1] [0x01] [packet 2] [0x02] ...
```

**Comment by zathras-crypto** [comment]
> As a side note: if we accomplish that the message field is not consensus critical, than it's not a strictly required for 0.0.10.

Whilst the field itself is not consensus critical, some of our discussed changes would invalidate the simple send in earlier clients...

> where the version field is split into 2x 8 bit fields, instead of 1x 16 bit field?

And I think we have a winner!  I never believed it was necessary to support >255 versions of a particular transaction (if a transaction goes through 250+ iterations it probably isn't the same transaction anymore and should have been a different type).  I think we 'steal' a byte from the version field (making version into an 8-bit field) and providing a new 8-bit field for some form of sequencing.

I'm worried we're blowing the scope out but why not aim high?!?!

> ```[0x00] [packet 1] [0x01] [packet 2] [0x02] ...```

So would you envisage ```[packet 1]```, ```[packet 2]``` etc including stuff like a message and HASH160 of recipient?

**Comment by dexX7** [comment]
> So would you envisage [packet 1], [packet 2] etc including stuff like a message and HASH160 of recipient?

That's a bit tricky.. both the message field as well as recipients are not really independent, but refer to a specific payload.

**Comment by zathras-crypto** [comment]
> That's a bit tricky.. both the message field as well as recipients are not really independent, but refer to a specific payload.

That's kind of what I meant mate - so:

```[packet 1 inc hash160 and message] [packet 2 inc hash160 and message]``` and so on...

**Comment by dexX7** [comment]
Right, which kinda brings us back to the beginning, but now we know this isn't necessarly a topic about multi-transactions, but rather about multi sub-payloads.

Assuming we'd add a version 1 for simple sends with messages, then how we'd tackle references?

We'd need to cover:

- [simple send]
- [simple send message]
- [simple send reference]
- [simple send reference message]
